asciidoc/modules/con_mg_changes-in-vertx-grpc.adoc

```diff
@@ -0,0 +1,8 @@
+[id="changes-in-vertx-grpc_{context}"]
+= Changes in {VertX} gRPC
+
+The following section describes the changes in {VertX} gRPC.
+
+== Removed custom {VertX} APIs for gRPC
+
+In the earlier releases of {VertX}, custom APIs for gRPC were available. To use this customized gRPC, a fork of the gRPC compiler had to be maintained. Since this was not easy to maintain, in {VertX} {4} the custom {VertX} APIs for gRPC have been removed.
```

Comment:
@vietj Who maintains the gRPC fork? If we maintain the fork, then giving this reason to the users for removing the APIs does not read well. Can you please help me rephrase the note?

Reply:
Actually I think the entire content is not correct now since we maintain a new customization of another gRPC compiler. CC @pmlopes what's your opinion?

Reply:
@pmlopes Can you please let us know your thoughts?

Reply:
@pmlopes ping

Reply:
@sangeetaraghu, what I wanted to say here was:


In vert.x 3.x the way to compile a grpc `.proto` descriptor involved a fork of the official grpc compiler maintained by the eclipse project. The maintenance was complex and required continuous rebase on the upstream code in order to keep up with the new development/features, plus had to be patched in order to support the vert.x specific generated code.

In 4.0 the project decided to abandon this fork and take the development into a more manageable manner. In 4.0 we have introduced a grpc compiler "plugin". The upstream compiler is plugin based, so this means that we are free from the upstream changes and users can now choose the upstream compiler version they wish.

Reply:
@pmlopes So do we want to tell the product users as well about the compiler plugin? What should the note say? 
That the custom APIs are not available, and to use the compiler plugin?

Reply:
Yes we can say that before we were using:

```xml
      <plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.vertx:protoc-gen-grpc-java:1.25.0:exe:${os.detected.classifier}</pluginArtifact>
        </configuration>
      </plugin>
```

And now the new configuration should include a plugin section like:

```xml
      <plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <version>0.6.1</version>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:${grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
          <protocPlugins>
            <protocPlugin>
              <id>vertx-grpc-protoc-plugin</id>
              <groupId>io.vertx</groupId>
              <artifactId>vertx-grpc-protoc-plugin</artifactId>
              <version>${vertx.version}</version>
              <mainClass>io.vertx.grpc.protoc.plugin.VertxGrpcGenerator</mainClass>
            </protocPlugin>
          </protocPlugins>
        </configuration>
      </plugin>
```

It is important to note that the line:

```
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:${grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
```

Is different between the 2 releases.

Reply:
the important bit is that the plugin in {4} is now the official `grpc` plugin, not a fork.

Reply:
@sangeetaraghu let me know if this is clear! Thanks!

Reply:
@pmlopes I am working on the note.. If I have questions can we please discuss on GChat/Discord?

Reply:
yes you can, there is a migration-guide section in the discord server

Reply:
@pmlopes I have rephrased as below. Does it read ok?

== New gRPC plugin

In earlier releases of {VertX}, to compile a gRPC `.proto` descriptor, you had to work with fork of the official gRPC compiler. This fork is maintained by the Eclipse project. Working with the fork was complex.

In previous releases, to work with gRPC, the following details were added to `pom.xml` file.

----
   <plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.vertx:protoc-gen-grpc-java:1.25.0:exe:${os.detected.classifier}</pluginArtifact>
        </configuration>
      </plugin>
----

In {VertX} {4}, a gRPC compiler plugin is available, which is independent of the offical gRPC fork. To work with the new gRPC plugin, add the following details to `pom.xml` file.

----
      <plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <version>0.6.1</version>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:${protoc.version}:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:${grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
          <protocPlugins>
            <protocPlugin>
              <id>vertx-grpc-protoc-plugin</id>
              <groupId>io.vertx</groupId>
              <artifactId>vertx-grpc-protoc-plugin</artifactId>
              <version>${vertx.version}</version>
              <mainClass>io.vertx.grpc.protoc.plugin.VertxGrpcGenerator</mainClass>
            </protocPlugin>
          </protocPlugins>
        </configuration>
      </plugin>
----

Reply:
Perhaps:

In {VertX} {4}, a gRPC compiler plugin is available, which is uses the **official** gRPC compiler. To work with the new gRPC plugin, add the following details to pom.xml file.

Reply:
As discussed, updating to:

In {VertX} {4}, a new gRPC compiler plugin is available. This plugin uses the official gRPC compiler instead of the fork. To work with the new gRPC plugin, add the following details to `pom.xml` file.

Reply:
Done. Thank you for the review @pmlopes